src/proxy/socks5.rs

```diff
@@ -185,40 +222,45 @@ async fn handle(mut oc: OutboundConnection, mut stream: TcpStream) -> Result<(),
     let mut atyp = [0u8];
     stream.read_exact(&mut atyp).await?;
 
-    let ip;
-
-    match atyp[0] {
+    let ip = match atyp[0] {
         0x01 => {
             let mut hostb = [0u8; 4];
             stream.read_exact(&mut hostb).await?;
-            ip = IpAddr::V4(hostb.into());
+            IpAddr::V4(hostb.into())
         }
         0x04 => {
             let mut hostb = [0u8; 16];
             stream.read_exact(&mut hostb).await?;
-            ip = IpAddr::V6(hostb.into());
+            IpAddr::V6(hostb.into())
         }
         0x03 => {
             let mut domain_length = [0u8];
             stream.read_exact(&mut domain_length).await?;
             let mut domain = vec![0u8; domain_length[0] as usize];
             stream.read_exact(&mut domain).await?;
-            // TODO: DNS lookup, if we want to integrate with HTTP-based apps without
-            // a DNS server.
-            let ds = std::str::from_utf8(&domain)?;
-            let Some(resolver) = &oc.pi.resolver else {
-                return Err(anyhow::anyhow!(
-                    "unsupported hostname lookup, requires DNS enabled"
+
+            let Ok(ds) = std::str::from_utf8(&domain) else {
+                return Err(SocksError::invalid_protocol(format!(
+                    "domain is not a valid utf8 string: {domain:?}"
+                )));
+            };
+            let Some(resolver) = &pi.resolver else {
+                return Err(SocksError::invalid_protocol(
+                    "unsupported hostname lookup, requires DNS enabled".to_string(),
                 ));
             };
 
-            ip = dns_lookup(resolver.clone(), remote_addr, ds).await?;
-            // oc.pi.resolver.lookup()
-            // oc.pi.lookup_service_or_query(ds)
-            // return Err(anyhow::anyhow!("unsupported host {ds:?}"));
+            match dns_lookup(resolver.clone(), remote_addr, ds).await {
+                Ok(ip) => ip,
+                Err(e) => {
+                    return Err(SocksError::HostUnreachable(e));
```

Comment:
Should this be like `UnknownHost` rather than unreachable?

Reply:
The only errors in the socks protocol are:

```
             o  X'01' general SOCKS server failure
             o  X'02' connection not allowed by ruleset
             o  X'03' Network unreachable
             o  X'04' Host unreachable
             o  X'05' Connection refused
             o  X'06' TTL expired
             o  X'07' Command not supported
             o  X'08' Address type not supported
```

Reply:
ah, yeah so the only real advantage would be around logging it more clearly maybe but at the end we would have to map this down to on of the above at which point we'd have this odd disconnect between what was logged and what was responded that would need mapping. Comment withdrawn